### PR TITLE
initial attempt at dockerizing cloudmapper

### PR DESCRIPTION
**Pull request by ngerstle**
This addition should make using a docker container pretty easy. 
I tend to prefer using docker-compose for most things, but I've documented what actions are necessary. 
It should be noted that the `demo` account name has a special meaning in the container, but it does here as well.
The should address #20 .

(also, thanks to my boss at Telenor Digital for saying yes to using the tool, repackaging it, and making an upstream PR)

### CODE REVIEW

**Comment by aboutte** [comment]
+1 for this.  I just tested it out and it works great!

**Comment by 0xdabbad00** [comment]
Thank you for this.  I don't have much Docker experience so I've asked someone else on my team to take a look.

**Comment by cybergoof** [comment]
How does the docker container get the aws configuration information?

**Comment by aboutte** [comment]
I set mine in docker-compose.yml:

```$ cat docker-compose.yml
version: "3.4"
services:
  cloudmapper:
    build:
      context: .
    environment:
      - AWS_ACCESS_KEY_ID=fdsfdasfdasfdsa
      - AWS_SECRET_ACCESS_KEY=fdsfdasfdafs
      - AWS_DEFAULT_REGION=us-gov-west-1
      - accountname=my-account-name
    ports:
      - "8000:8000"```

**Comment by ngerstle** [comment]
The environment section in AWS compose will pull in the noted environment variables if defined in the shell environment, which is what happens if you use aws-vault or run it on an ec2 instance:
```
environment:
     - AWS_ACCESS_KEY_ID
     - AWS_SECRET_ACCESS_KEY
     - AWS_SESSION_TOKEN
     - AWS_SECURITY_TOKEN
     - AWS_DEFAULT_REGION=eu-west-1
     - accountname=accountname
```

Thus, if it isn't defined in your shell (check with `$ printenv | grep -i aws`), you could run:
```
$ export AWS_ACCESS_KEY=Ifadfdasfsadfasdf
$ export AWS_SECRET_ACCESS_KEY=asdasdasd
...
$ docker-compose up
```
or 
```
$ AWS_ACCESS_KEY=Ifadfdasfsadfasdf   AWS_SECRET_ACCESS_KEY=asdasdasd  docker-compose up
```

but all that leaves your keys in your shell history, which isn't great, or in a likely to be committed file, as above by @aboutte  (which does work).

**Ideally, I'd use aws-vault:**
```
 aws-vault exec accountname -- docker-compose up
```

**Comment by cybergoof** [comment]
I would recommend splitting it up.  The keys are only needed to grab the data, which is a one and done thing.  I would think about breaking this up. One docker-compose that pulls the information, prepares, then stores in a docker volume (then dies).  Then another compose file that runs the webserver, pulling from that docker volume.  That way you don't have anything running that has the keys.

**Comment by ngerstle** [comment]
@cybergoof: I could split it up into two different docker images (the collection/preparation of data vs serving it up) that share a volume, if preferred, but I don't see any particular advantage with respect to keys- either way you're going to be running a python server in docker on likely the same host you have your keys.
It would make it easier to grab the data.json and copy the rest of web/ to be served by a basic server container (something with a tiny statically compiled binary, nginx/apache so authentication can be more easily integrated, whatever).

Alternatively, I could add a proper switch (not just `accountname="demo"`) to skip the initial data collection, so you could run `COLLECT=True docker-compose up`, copy the result, then `COLLECT=False docker-compose up`  or something...

(sorry if that was a bit ramblely- I need to close my laptop for the weekend :smile: )

**Comment by cybergoof** [comment]
True.  But the docker container is really great at deploying for multiple people to access.  You don't want your keys in there.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+services:
+  cloudmapper:
+    build:
+      context: .
+    environment:
+      - AWS_ACCESS_KEY_ID
+      - AWS_SECRET_ACCESS_KEY
+      - AWS_SESSION_TOKEN
+      - AWS_SECURITY_TOKEN
+      - AWS_DEFAULT_REGION=eu-west-1
```
Comment (stephengroat) [inline docker-compose.yml]: why override `AWS_DEFAULT_REGION`? i use `aws-vault`, which defaultly sets the region for me and was wondering if that could be respected in `docker-compose`

**Comment by FernandoMiguel** [comment]
Oh i missed this PR
I've opened another 
https://github.com/duo-labs/cloudmapper/pull/54

File: docker-compose.yml (hunk 1 above)
Comment (ngerstle) [reply docker-compose.yml]: I had issues with `AWS_DEFAULT_REGION`- not being set. It doesn't seem to matter, as the tool pulls a list of valid regions, but in order to pull a list of regions, a default region needs to be set.

**Comment by ngerstle** [comment]
I've integrated parts of @FernandoMiguel Dockerfile to keep it slim, and use Alpine rather than Debian, but kept the docker-compose file and entrypoint.sh.

File: entrypoint.sh [hunk 2]
```diff
@@ -1,10 +1,11 @@
...
-if [ "$accountname" != "demo" ] 
+if [ "$ACCOUNT" != "demo" ] 
 then
-	./collect_data.sh --account $accountname
+  pipenv run python cloudmapper.py gather --account-name $ACCOUNT
 else
 	cp config.json.demo config.json
 fi
-pipenv run python cloudmapper.py prepare --account $accountname
+pipenv run python cloudmapper.py prepare --account $ACCOUNT
```
Comment (FernandoMiguel) [inline entrypoint.sh]: needs to be --account-name
you can see it here 
https://github.com/duo-labs/cloudmapper/pull/54/files

Comment (ngerstle) [reply entrypoint.sh]: Will fix- give me a minute

**Comment by ngerstle** [comment]
@FernandoMiguel 
I prefer a the Dockerfile I've committed for several (admittedly small) style reasons:
- I prefer separate out the `CMD` block into a distinct script, even if it is a short one. Putting the command(s) to be run in exec form, rather than shell, is recommended in order to avoid potential zombie pids or other issues as well.
- Placement of the code in `/opt/cloudmapper`, rather than `/cloudmapper` (not perfect, but more in keeping with the file system hierarchy standard)
- Placing the `CMD` at the end; and general order of directives in the Dockerfile;
- Use of pipenv- not strictly necessary, given the scoping by using docker in the first place, but I try to do all development of python using pipenv rather than pip where possible.

Nothing major, just small stuff.

**Comment by FernandoMiguel** [comment]
One of the reasons i place CMD so high is for cache busting reasons.
CMD won't change often, so no need to lose one cache layer.

**Comment by 0xdabbad00** [comment]
Thanks @FernandoMiguel and @ngerstle for your contributions on this!

**Pull request merged**